src/category_theory/sigma/basic.lean

```diff
@@ -0,0 +1,253 @@
+`desc F`.
+-/
+def desc_uniq (q : (Σ i, C i) ⥤ D) (h : Π i, incl i ⋙ q ≅ F i) : q ≅ desc F :=
+nat_iso.of_components (λ ⟨i, X⟩, (h i).app X) $
+  by { rintro ⟨i, X⟩ ⟨_, _⟩ ⟨_, _, Y, f⟩, apply (h i).hom.naturality f }
+
+@[simp]
+lemma desc_uniq_hom_app (q : (Σ i, C i) ⥤ D) (h : Π i, incl i ⋙ q ≅ F i) (i : I) (X : C i) :
+  (desc_uniq F q h).hom.app ⟨i, X⟩ = (h i).hom.app X :=
+rfl
+
+@[simp]
+lemma desc_uniq_inv_app (q : (Σ i, C i) ⥤ D) (h : Π i, incl i ⋙ q ≅ F i) (i : I) (X : C i) :
+  (desc_uniq F q h).inv.app ⟨i, X⟩ = (h i).inv.app X :=
+rfl
+
+/--
+If `q₁` and `q₂` when restricted to each subcategory `C i` agree, then `q₁` and `q₂` are isomorphic.
+-/
+@[simps]
+def nat_iso {q₁ q₂ : (Σ i, C i) ⥤ D} (h : Π i, incl i ⋙ q₁ ≅ incl i ⋙ q₂) :
+  q₁ ≅ q₂ :=
+{ hom := nat_trans (λ i, (h i).hom),
+  inv := nat_trans (λ i, (h i).inv) }
+
+end
+
+section
+
+variables (C) {J : Type w₂} (g : J → I)
+
+/-- A function `J → I` induces a functor `Σ j, C (g j) ⥤ Σ i, C i`. -/
+def map : (Σ (j : J), C (g j)) ⥤ (Σ (i : I), C i) :=
+desc (λ j, incl (g j))
+
+@[simp] lemma map_obj (j : J) (X : C (g j)) : (sigma.map C g).obj ⟨j, X⟩ = ⟨g j, X⟩ := rfl
+@[simp] lemma map_map {j : J} {X Y : C (g j)} (f : X ⟶ Y) :
+  (sigma.map C g).map (sigma_hom.mk f) = sigma_hom.mk f :=
+rfl
+
+/--
+The functor `sigma.map C g` restricted to the subcategory `C j` acts as the inclusion of `g j`.
+-/
+@[simps {rhs_md := semireducible, simp_rhs := tt}]
+def incl_comp_map (j : J) : incl j ⋙ map C g ≅ incl (g j) := iso.refl _
+
+variable (I)
+
+/-- The functor `sigma.map` applied to the identity function is just the identity functor. -/
+@[simps {rhs_md := semireducible, simp_rhs := tt}]
+def map_id : map C (id : I → I) ≅ 𝟭 (Σ i, C i) :=
+nat_iso (λ i, nat_iso.of_components (λ X, iso.refl _) (by tidy))
+
+variables {I} {K : Type w₃}
+
+/-- The functor `sigma.map` applied to a composition is a composition of functors. -/
+@[simps {rhs_md := semireducible, simp_rhs := tt}]
+def map_comp (f : K → J) (g : J → I) : map (C ∘ g) f ⋙ (map C g : _) ≅ map C (g ∘ f) :=
+desc_uniq _ _ $ λ k,
+  (iso_whisker_right (incl_comp_map (C ∘ g) f k) (map C g : _) : _) ≪≫ incl_comp_map _ _ _
+
+end
+
+namespace functor
+
+variables {C}
+variables {D : I → Type u₁} [∀ i, category.{v₁} (D i)]
+
+/--
+Assemble an `I`-indexed family of functors into a functor between the sigma types.
+-/
+def sigma (F : Π i, C i ⥤ D i) : (Σ i, C i) ⥤ (Σ i, D i) :=
+desc (λ i, F i ⋙ incl i)
+
+end functor
+
+namespace nat_trans
+
+variables {C}
+variables {D : I → Type u₁} [∀ i, category.{v₁} (D i)]
+variables {F G : Π i, C i ⥤ D i}
+
+/--
+Assemble an `I`-indexed family of natural transformations into a single natural transformation.
+-/
+def sigma (α : Π i, F i ⟶ G i) : functor.sigma F ⟶ functor.sigma G :=
+{ app := λ f, sigma_hom.mk ((α f.1).app _),
+  naturality' :=
+  begin
+    rintro ⟨i, X⟩ ⟨_, _⟩ ⟨_, _, Y, f⟩,
+    change sigma_hom.mk _ = sigma_hom.mk _,
+    rw (α i).naturality,
```

Comment:
Can you use the equation compiler here and make this a one-liner? I'm sorry I'm not trying this myself :-(

Reply:
I gave it a go and I couldn't get it to work, feel free to have a go yourself!

Reply:
Taking a look again, I think it's because there's a difference between using angle brackets in a lambda and using the equation compiler to build a def directly. For the `app` before, using a one-line definition seems to make `simps` give worse lemmas, but I'll make the change anyway

Reply:
Amended, thanks